### PR TITLE
8311932: Suboptimal compiled code of nested loop over memory segment

### PR DESCRIPTION
**Pull request by rwestrel**
To enable the elimination of long range checks, the loop that contains
the range checks is transformed into a loop nest and the range checks
are changed to operate on int values computed before the loop is
entered. This causes extra overhead out of loop and once, the range
checks are eliminated, can only pay off if the loop is executed for
long enough. This change disable the transformation if the trip count
computed from profile data is too low. This came up with a
MemorySegment API micro benchmarks and improves performance
significantly.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311932](https://bugs.openjdk.org/browse/JDK-8311932): Suboptimal compiled code of nested loop over memory segment (**Bug** - P2)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [a221320e](https://git.openjdk.org/jdk/pull/16650/files/a221320e82b61578b5be83d675a37aa5a9599f27)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16650/head:pull/16650` \
`$ git checkout pull/16650`

Update a local copy of the PR: \
`$ git checkout pull/16650` \
`$ git pull https://git.openjdk.org/jdk.git pull/16650/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16650`

View PR using the GUI difftool: \
`$ git pr show -t 16650`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16650.diff">https://git.openjdk.org/jdk/pull/16650.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16650#issuecomment-1809807337)

### CODE REVIEW

**Comment by TobiHartmann** [comment]
`TestLongRangeChecks.java` fails with `-ea -esa -XX:CompileThreshold=100 -XX:+UnlockExperimentalVMOptions -server -XX:-TieredCompilation`:

```
1) Method "public static void compiler.c2.irTests.TestLongRangeChecks.testStridePosScalePosInIntLoopShortLoop1(int,int,long,long)" - [Failed IR rules: 2]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#COUNTED_LOOP#_", "1"}, applyIfPlatform={}, applyIfPlatformOr={}, failOn={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(CountedLoop\\b.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!
   * @IR rule 2: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={}, applyIfPlatform={}, applyIfPlatformOr={}, failOn={"_#LOOP#_"}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - failOn: Graph contains forbidden nodes:
         * Constraint 1: "(\\d+(\\s){2}(Loop.*)+(\\s){2}===.*)"
           - Matched forbidden node:
             * 173  Loop  === 173 167 170  [[ 173 420 174 ]] inner 

2) Method "public static void compiler.c2.irTests.TestLongRangeChecks.testStridePosScalePosShortLoop(long,long,long,long)" - [Failed IR rules: 2]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#LONG_COUNTED_LOOP#_", "1"}, applyIfPlatform={}, applyIfPlatformOr={}, failOn={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(LongCountedLoop\\b.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!
   * @IR rule 2: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={}, applyIfPlatform={}, applyIfPlatformOr={}, failOn={"_#COUNTED_LOOP#_", "_#LOOP#_"}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - failOn: Graph contains forbidden nodes:
         * Constraint 2: "(\\d+(\\s){2}(Loop.*)+(\\s){2}===.*)"
           - Matched forbidden node:
             * 170  Loop  === 170 165 167  [[ 170 403 171 ]] inner 

```

**Comment by rwestrel** [comment]
> `TestLongRangeChecks.java` fails with `-ea -esa -XX:CompileThreshold=100 -XX:+UnlockExperimentalVMOptions -server -XX:-TieredCompilation`:

Thanks for running testing. Should be fixed now.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. I've re-submitted testing and report back once it passed.

File: test/hotspot/jtreg/compiler/c2/irTests/TestLongRangeChecks.java [hunk 1]
```diff
@@ -38,9 +38,9 @@
 
 public class TestLongRangeChecks {
     public static void main(String[] args) {
-        TestFramework.runWithFlags("-XX:-UseCountedLoopSafepoints");
-        TestFramework.runWithFlags("-XX:+UseCountedLoopSafepoints", "-XX:LoopStripMiningIter=1");
-        TestFramework.runWithFlags("-XX:+UseCountedLoopSafepoints", "-XX:LoopStripMiningIter=1000");
+        TestFramework.runWithFlags("-XX:+TieredCompilation", "-XX:-UseCountedLoopSafepoints", "-XX:LoopUnrollLimit=0");
```
[the remark this answers is not in the document]
Comment (rwestrel) [reply test/hotspot/jtreg/compiler/c2/irTests/TestLongRangeChecks.java:41]: Done in new commit.

**Comment by rwestrel** [comment]
> Looks good!

Thanks for reviewing and for the suggestions.

**Review by TobiHartmann — APPROVED** [review]
All tests passed.

File: src/hotspot/share/opto/loopnode.cpp [hunk 2]
```diff
@@ -855,6 +855,28 @@ bool PhaseIdealLoop::create_loop_nest(IdealLoopTree* loop, Node_List &old_new) {
     // not a loop after all
     return false;
   }
+
+  if (range_checks.size() > 0) {
+    // This transformation requires peeling one iteration. Also, if it has range checks and they are eliminated by
+    // Loop Predication, then 2 Hoisted Check Predicates are added for one range check. Finally, transforming a long range check requires
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopnode.cpp:861]: Just noticed that the line wraps are off now - maybe you want to re-wrap these lines. But that's just a minor detail.

**Review by chhagedorn — APPROVED** [review]
That looks good, thanks for the update.

**Comment by rwestrel** [comment]
@TobiHartmann @chhagedorn thanks for the reviews

**Pull request closed without merging**